// baobab-react/dist-modules/higher-order.js
/**
 * Baobab-React Higher Order Component
 * ====================================
 *
 * ES6 higher order component to enchance one's component.
 */
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x2, _x3, _x4) { var _again = true; _function: while (_again) { var object = _x2, property = _x3, receiver = _x4; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x2 = parent; _x3 = property; _x4 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

exports.root = root;
exports.branch = branch;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _baobab = require('baobab');

var _baobab2 = _interopRequireDefault(_baobab);

var _utilsHelpersJs = require('./utils/helpers.js');

var _utilsPropTypesJs = require('./utils/prop-types.js');

var _utilsPropTypesJs2 = _interopRequireDefault(_utilsPropTypesJs);

var makeError = _baobab2['default'].helpers.makeError;

/**
 * Root component
 */

function root(Component, tree) {
  if (!(tree instanceof _baobab2['default'])) throw makeError('baobab-react:higher-order.root: given tree is not a Baobab.', { target: tree });

  var componentDisplayName = Component.name || Component.displayName || 'Component';

  var ComposedComponent = (function (_React$Component) {
    _inherits(ComposedComponent, _React$Component);

    function ComposedComponent() {
      _classCallCheck(this, ComposedComponent);

      _get(Object.getPrototypeOf(ComposedComponent.prototype), 'constructor', this).apply(this, arguments);
    }

    _createClass(ComposedComponent, [{
      key: 'getChildContext',

      // Handling child context
      value: function getChildContext() {
        return {
          tree: tree
        };
      }

      // Render shim
    }, {
      key: 'render',
      value: function render() {
        return _react2['default'].createElement(Component, this.props);
      }
    }], [{
      key: 'displayName',
      value: 'Rooted' + componentDisplayName,
      enumerable: true
    }, {
      key: 'childContextTypes',
      value: {
        tree: _utilsPropTypesJs2['default'].baobab
      },
      enumerable: true
    }]);

    return ComposedComponent;
  })(_react2['default'].Component);

  return ComposedComponent;
}

/**
 * Branch component
 */

function branch(Component) {
  var mapping = arguments.length <= 1 || arguments[1] === undefined ? null : arguments[1];

  var componentDisplayName = Component.name || Component.displayName || 'Component';

  var ComposedComponent = (function (_React$Component2) {
    _inherits(ComposedComponent, _React$Component2);

    _createClass(ComposedComponent, [{
      key: 'getChildContext',

      // Passing the component's cursors through context
      value: function getChildContext() {
        return this.cursors ? {
          cursors: this.cursors
        } : {};
      }

      // Building initial state
    }], [{
      key: 'displayName',
      value: 'Branched' + componentDisplayName,
      enumerable: true
    }, {
      key: 'contextTypes',
      value: {
        tree: _utilsPropTypesJs2['default'].baobab
      },
      enumerable: true
    }, {
      key: 'childContextTypes',
      value: {
        cursors: _utilsPropTypesJs2['default'].cursors
      },
      enumerable: true
    }]);

    function ComposedComponent(props, context) {
      _classCallCheck(this, ComposedComponent);

      _get(Object.getPrototypeOf(ComposedComponent.prototype), 'constructor', this).call(this, props, context);

      if (mapping.cursors) {
        var solvedMapping = (0, _utilsHelpersJs.solveMapping)(mapping.cursors, props, context);

        if (!solvedMapping) throw makeError('baobab-react:higher-order.branch: given cursors mapping is invalid (check the "' + displayName + '" component).', { mapping: solvedMapping });

        // Creating the watcher
        this.watcher = this.context.tree.watch(solvedMapping);
        this.cursors = this.watcher.getCursors();
        this.state = this.watcher.get();
      }
    }

    // On component mount

    _createClass(ComposedComponent, [{
      key: 'componentWillMount',
      value: function componentWillMount() {
        if (!this.watcher) return;

        var handler = (function () {
          if (this.watcher) this.setState(this.watcher.get());
        }).bind(this);

        this.watcher.on('update', handler);
      }

      // Render shim
    }, {
      key: 'render',
      value: function render() {
        var tree = this.context.tree,
            suppl = {};

        // Binding actions if any
        if (mapping.actions) {
          suppl.actions = {};

          Object.keys(mapping.actions).forEach(function (k) {
            suppl.actions[k] = mapping.actions[k].bind(tree, tree);
          });
        }

        return _react2['default'].createElement(Component, _extends({}, this.props, suppl, this.state));
      }

      // On component unmount
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        if (!this.watcher) return;

        // Releasing watcher
        this.watcher.release();
        this.watcher = null;
      }

      // On new props
    }, {
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(props) {
        if (!this.watcher || !mapping.cursors) return;

        var solvedMapping = (0, _utilsHelpersJs.solveMapping)(mapping.cursors, props, this.context);

        if (!solvedMapping) throw makeError('baobab-react:higher-order.branch: given mapping is invalid (check the "' + displayName + '" component).', { mapping: solvedMapping });

        // Refreshing the watcher
        this.watcher.refresh(solvedMapping);
        this.cursors = this.watcher.getCursors();
        this.setState(this.watcher.get());
      }
    }]);

    return ComposedComponent;
  })(_react2['default'].Component);

  return ComposedComponent;
}
